### PR TITLE
reputation v2

### PR DESCRIPTION
**Pull request by Stumble**

### CODE REVIEW

**Comment by Stumble** [comment]
TODO: change update6's blockheight after pr is reviewed.

File: x/reputation/manager.go [hunk 1]
```diff
@@ -115,6 +145,10 @@ func (rep ReputationManager) OnStakeIn(ctx sdk.Context,
 // OnStakeOut - on @p username stakeout @p amount
 func (rep ReputationManager) OnStakeOut(ctx sdk.Context,
 	username types.AccountKey, amount types.Coin) {
+	// after upgrade6, no-op on stackout.
+	if ctx.BlockHeight() >= types.BlockchainUpgrade1Update6Height {
```
Comment (xiaoxiaff) [inline x/reputation/manager.go]: if no deduction on stake out then stake in can earn free score?

File: x/reputation/manager.go [hunk 2]
```diff
@@ -186,26 +243,50 @@ func (rep ReputationManager) GetCurrentRound(ctx sdk.Context) (int64, sdk.Error)
...
 	if err != nil {
 		return err
 	}
-	handler.ExportToFile(file)
+	v1RepStore := model.NewReputationStore(ctx.KVStore(rep.v1key), param.BestContentIndexN)
+	v1handler := model.NewReputation(v1RepStore)
+	// Update6, if a user does not donate after update6, his reputation is reset to 0.
+	if ctx.BlockHeight() >= types.BlockchainUpgrade1Update6Height {
+		repv2 := rep.getHandlerV2(ctx)
+		if err != nil {
```
Comment (ZhimaoL) [inline x/reputation/manager.go]: same?

File: x/reputation/repv2/store.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+// data necessary for reputation system. Also, it takes the problem, that the underlying
+// kv store may be an iavl, into account so the number of keys will have a large
+// impact on performance into account, by trying to minimize the number of keys.
+// we choosed amino json as serializer because it is deterministic.
+
+var (
+	KeySeparator           = byte('/')
+	repUserMetaPrefix      = []byte{0x00}
+	repRoundMetaPrefix     = []byte{0x01}
+	repRoundPostMetaPrefix = []byte{0x02}
```
Comment (xiaoxiaff) [inline x/reputation/repv2/store.go]: deprecated

File: x/reputation/repv2/store_test.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+func (suite *StoreTestSuite) SetupTest() {
+	suite.mockDB = internal.NewMockStore()
+}
+
+func (suite *StoreTestSuite) TestPrefix() {
+	suite.Equal(append(repUserMetaPrefix, []byte("qwe")...), getUserMetaKey("qwe"))
+	suite.Equal(append(repRoundMetaPrefix, []byte("3")...), getRoundMetaKey(3))
+	suite.Equal(append(repRoundPostMetaPrefix,
```
Comment (xiaoxiaff) [inline x/reputation/repv2/store_test.go]: deprecated

**Pull request merged**